# Chess Dataset Analysis
## Installing necessary libraries and importing the dataset (Step 1)

In [9]:
import pandas as pd

df = pd.read_csv('datasets/games.csv')


**Chess Games dataset** contains more than 20,000 rows and 16 columns.

## Analysis Questions and Hypotheses (Step 2)

### Q1 - Does the higher-rated player win more often?
**Type:** Correlation between variables  
**Hypothesis:** Players with a higher rating than their opponent will more games. The winrate of the higher-rated player will be way above 50%

### Q2 - Does playing as white give a bigger advantage?
**Type:** Comparison between groups  
**Hypothesis:** White wins more games than black, since they have an advantage of playing the first move. White's win rate will be above 50% across all games

### Q3 - Do higher-rated games last longer?
**Type:** Correlation between variables  
**Hypothesis:** Games between higher-rated players will have more moves on average, as stronger players are less likely to blunder early and tend to play more positionally, leading to the endgame

### Q4 - Does the choice of opening affects the outcome?
**Type:** Impact of one variable on another  
**Hypothesis:** Certain openings will have a statistically higher win rate for white or black. Aggressive openings (such as Sicilian Defense or King's Gambit ) will result in more wins by checkmate or resignation of the opponent compared to positional ones.

### Q5 - Are rated games played differently from unrated games?
**Type:** Comparison between groups  
**Hypothesis:** Rated games will last longer and end in resignation or checkmate more often, while unrated ones will have more losses due to timeouts, meaning that players tend to take rated games more seriously.


## Exploring the data (Step 3)

In [10]:

df.head()

,id,rated,created_at,last_move_at,turns,victory_status,winner,increment_code,white_id,white_rating,black_id,black_rating,moves,opening_eco,opening_name,opening_ply
0,TZJHLljE,False,1.504210e+12,1.504210e+12,13,outoftime,white,15+2,bourgris,1500,a-00,1191,d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...,D10,Slav Defense: Exchange Variation,5
1,l1NXvwaE,True,1.504130e+12,1.504130e+12,16,resign,black,5+10,a-00,1322,skinnerua,1261,d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...,B00,Nimzowitsch Defense: Kennedy Variation,4
2,mIICvQHh,True,1.504130e+12,1.504130e+12,61,mate,white,5+10,ischia,1496,a-00,1500,e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...,C20,King's Pawn Game: Leonardis Variation,3
3,kWKvrqYL,True,1.504110e+12,1.504110e+12,61,mate,white,20+0,daniamurashov,1439,adivanov2009,1454,d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...,D02,Queen's Pawn Game: Zukertort Variation,3
4,9tXo1AUZ,True,1.504030e+12,1.504030e+12,95,mate,white,30+3,nik221107,1523,adivanov2009,1469,e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...,C41,Philidor Defense,5


In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  str    
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  str    
 6   winner          20058 non-null  str    
 7   increment_code  20058 non-null  str    
 8   white_id        20058 non-null  str    
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  str    
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  str    
 13  opening_eco     20058 non-null  str    
 14  opening_name    20058 non-null  str    
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), str(9)
memory usage: 2.3 MB


In [12]:
df.describe()

,created_at,last_move_at,turns,white_rating,black_rating,opening_ply
count,2.005800e+04,2.005800e+04,20058.000000,20058.000000,20058.000000,20058.000000
mean,1.483617e+12,1.483618e+12,60.465999,1596.631868,1588.831987,4.816981
std,2.850151e+10,2.850140e+10,33.570585,291.253376,291.036126,2.797152
min,1.376772e+12,1.376772e+12,1.000000,784.000000,789.000000,1.000000
25%,1.477548e+12,1.477548e+12,37.000000,1398.000000,1391.000000,3.000000
50%,1.496010e+12,1.496010e+12,55.000000,1567.000000,1562.000000,4.000000
75%,1.503170e+12,1.503170e+12,79.000000,1793.000000,1784.000000,6.000000
max,1.504493e+12,1.504494e+12,349.000000,2700.000000,2723.000000,28.000000


## Missing Values and Key Observations (Step 3)

**Data Types and General Observations:**
- `created_at` and `last_move_at` column doesn't have a correct time format
- `moves` is a very long text column
- `id` has only 19113 unique values out of all 20058 rows
- both `white_rating` and `black_rating` columns have a wide range, having both beginners (below 800 elo) and chess masters (above 2000 elo)

## Clean and Transform the Data (Step 4)

In [13]:
df = df.drop_duplicates() # cleaning dataset from duplicates

# dataset had 429 duplicates

In [ ]:
# Removing potential outliers for future analysis

df = df[df['turns'] > 1] # deleting the games where the opponent quit/resigned 
 # before the game even started

# dataset had 17 games with just 1 turn

19629
19612


In [ ]:
# converting the data types of 'created_at' and 'last_move_at' to correct DateTime

df['created_at'] = pd.to_datetime(df['created_at'], unit='ms')
df['last_move_at'] = pd.to_datetime(df['last_move_at'], unit='ms')

df[['created_at', 'last_move_at']].head()

# Now these 2 columns have correct date format

,created_at,last_move_at
0,2017-08-31 20:06:40,2017-08-31 20:06:40
1,2017-08-30 21:53:20,2017-08-30 21:53:20
2,2017-08-30 21:53:20,2017-08-30 21:53:20
3,2017-08-30 16:20:00,2017-08-30 16:20:00
4,2017-08-29 18:06:40,2017-08-29 18:06:40


In [17]:
df['victory_status'] = df['victory_status'].str.lower().str.strip()
df['victory_status'] = df['victory_status'].str.lower().str.strip()

# making sure that categories are standardised - all lowercase and without any whitespaces

In [18]:
# renaming the columns for more clarity during the analysis

df = df.rename(columns={
    'white_rating': 'rating_white',
    'black_rating': 'rating_black',
    'opening_ply': 'opening_moves',
    'opening_eco': 'opening_eco_code',
})

In [ ]:
# New Features

df['rating_diff'] = df['rating_white'] - df['rating_black'] # rating difference between opponents
df['avg_rating'] = ((df['rating_white'] + df['rating_black']) / 2).round(0).astype(int)

,id,rated,created_at,last_move_at,turns,victory_status,winner,increment_code,white_id,rating_white,black_id,rating_black,moves,opening_eco_code,opening_name,opening_moves,rating_diff,avg_rating
0,TZJHLljE,False,2017-08-31 20:06:40,2017-08-31 20:06:40,13,outoftime,white,15+2,bourgris,1500,a-00,1191,d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...,D10,Slav Defense: Exchange Variation,5,309,1346
1,l1NXvwaE,True,2017-08-30 21:53:20,2017-08-30 21:53:20,16,resign,black,5+10,a-00,1322,skinnerua,1261,d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...,B00,Nimzowitsch Defense: Kennedy Variation,4,61,1292
2,mIICvQHh,True,2017-08-30 21:53:20,2017-08-30 21:53:20,61,mate,white,5+10,ischia,1496,a-00,1500,e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...,C20,King's Pawn Game: Leonardis Variation,3,-4,1498
3,kWKvrqYL,True,2017-08-30 16:20:00,2017-08-30 16:20:00,61,mate,white,20+0,daniamurashov,1439,adivanov2009,1454,d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...,D02,Queen's Pawn Game: Zukertort Variation,3,-15,1446
4,9tXo1AUZ,True,2017-08-29 18:06:40,2017-08-29 18:06:40,95,mate,white,30+3,nik221107,1523,adivanov2009,1469,e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...,C41,Philidor Defense,5,54,1496


In [26]:
# Game length

df['game_length'] = pd.cut(
    df['turns'],
    bins=[0, 20, 40, 60, 100, 999],
    labels=['Very Short', 'Short', 'Medium', 'Long', 'Very Long']
)

In [27]:
# Binary target for future ML Model

df['white_won'] = (df['winner'] == 'white').astype(int)

In [31]:
# Encoding categories varialbes for ML Model

from sklearn.preprocessing import LabelEncoder

le_victory_status = LabelEncoder()
le_winner = LabelEncoder()

df['victory_status_enc'] = le_victory_status.fit_transform(df['victory_status'])
df['winner_enc'] = le_winner.fit_transform(df['winner'])


## Clean Dataset ready for further analysis/ML/AI making

In [32]:
df.head()

,id,rated,created_at,last_move_at,turns,victory_status,winner,increment_code,white_id,rating_white,...,moves,opening_eco_code,opening_name,opening_moves,rating_diff,avg_rating,game_length,white_won,victory_status_enc,winner_enc
0,TZJHLljE,False,2017-08-31 20:06:40,2017-08-31 20:06:40,13,outoftime,white,15+2,bourgris,1500,...,d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...,D10,Slav Defense: Exchange Variation,5,309,1346,Very Short,1,2,2
1,l1NXvwaE,True,2017-08-30 21:53:20,2017-08-30 21:53:20,16,resign,black,5+10,a-00,1322,...,d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...,B00,Nimzowitsch Defense: Kennedy Variation,4,61,1292,Very Short,0,3,0
2,mIICvQHh,True,2017-08-30 21:53:20,2017-08-30 21:53:20,61,mate,white,5+10,ischia,1496,...,e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...,C20,King's Pawn Game: Leonardis Variation,3,-4,1498,Long,1,1,2
3,kWKvrqYL,True,2017-08-30 16:20:00,2017-08-30 16:20:00,61,mate,white,20+0,daniamurashov,1439,...,d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...,D02,Queen's Pawn Game: Zukertort Variation,3,-15,1446,Long,1,1,2
4,9tXo1AUZ,True,2017-08-29 18:06:40,2017-08-29 18:06:40,95,mate,white,30+3,nik221107,1523,...,e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...,C41,Philidor Defense,5,54,1496,Long,1,1,2


In [35]:
df.to_csv('datasets/chess_games_clean.csv', index=False) # making a new clean dataset